presto-hive/src/main/java/com/facebook/presto/hive/statistics/ParquetQuickStatsBuilder.java

```diff
@@ -0,0 +1,346 @@
+import com.facebook.presto.hive.metastore.Partition;
+import com.facebook.presto.hive.metastore.SemiTransactionalHiveMetastore;
+import com.facebook.presto.hive.metastore.StorageFormat;
+import com.facebook.presto.hive.metastore.Table;
+import com.facebook.presto.parquet.ParquetDataSource;
+import com.facebook.presto.parquet.cache.ParquetFileMetadata;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.SchemaTableName;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.Slice;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.Path;
+import org.apache.parquet.column.statistics.DoubleStatistics;
+import org.apache.parquet.column.statistics.FloatStatistics;
+import org.apache.parquet.column.statistics.IntStatistics;
+import org.apache.parquet.column.statistics.LongStatistics;
+import org.apache.parquet.column.statistics.Statistics;
+import org.apache.parquet.hadoop.metadata.BlockMetaData;
+import org.apache.parquet.hadoop.metadata.ColumnChunkMetaData;
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+import org.apache.parquet.schema.LogicalTypeAnnotation.DateLogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.LogicalTypeAnnotationVisitor;
+import org.apache.parquet.schema.LogicalTypeAnnotation.TimeLogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+
+import java.time.LocalDate;
+import java.time.chrono.ChronoLocalDate;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.hive.CacheQuota.NO_CACHE_CONSTRAINTS;
+import static com.facebook.presto.hive.HivePartition.UNPARTITIONED_ID;
+import static com.facebook.presto.hive.HiveSessionProperties.getReadNullMaskedParquetEncryptedValue;
+import static com.facebook.presto.hive.parquet.HdfsParquetDataSource.buildHdfsParquetDataSource;
+import static com.facebook.presto.hive.parquet.ParquetPageSourceFactory.createDecryptor;
+import static com.facebook.presto.parquet.cache.MetadataReader.readFooter;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class ParquetQuickStatsBuilder
+        implements QuickStatsBuilderInterface
+{
+    public static final Logger log = Logger.get(ParquetQuickStatsBuilder.class);
+    private static final ExecutorService footerFetchExecutor = Executors.newCachedThreadPool();
```

Comment:
I understand wanting to control size but the `newCachedThreadPool()` with no parameters basically grows unbounded as you throw work at it to prevent queuing. You might get more concurrency due to many IO operations but you can run into high context switch overheads or too many threads if ulimits are not configured high enough. I would at least use a thread pool where you can cap the threads.

I was going to suggest just using the common FJ pool. The common pool is capped at 2x core count which results in queuing but limits overheads. Since it's shared though you may run into queuing issues. Keeping separate executors is OK, but I would at least looking at capping the thread count

Reply:
Okay, any suggestions on sane defaults for the max threads for the bounded executors in ParquetQuickStatsBuilder and QuickStatsProvider ?

Reply:
I think there should be a cap, but unsure what would be reasonable. Just spitting out my thoughts here

A single executor runnable can't be suspended when an IO request is made, thus if you have, say 400 outstanding requests and an executor with 2 threads, you'll only ever get a maximum concurrency of 2 requests. In reality, a properly constructed IO system should be able to submit all 400 requests using the 2 threads concurrently and return the results for each once the IO is finished, without ever blocking a thread. Unfortunately, not something we have available in Java using the `FileSystem` APIs that make the IO requests.

Since most of the requests are IO-based and not CPU-based, increasing the executor size past the thread count of the machine **is** likely to have tangible benefit from my understanding. The only worry would be memory consumption of the thread's stack space in the case of hundreds (or potentially thousands) or threads. My guess is that the cap should probably be some function of the JVM memory, not necessarily the available cores on the machine if you assume the work these threads do is IO-bound, not CPU bound.

I believe the default stack size is 1M, meaning 1024 threads would consume _up to_ 1GiB of memory. Maybe scaling the thread count to something like consuming up to 10/15% of max memory by default might be reasonable?  e.g. if the coordinator has 10G heap, then you would allow 10% of that memory to be consumed by quick stats threads IF they reach maximum stack size. Most of the time this would be much smaller as thread stack minimum size is something like 16KiB.

Example for ParquetQuickStatsBuilder

```java
private static final ExecutorService footerFetchExecutor = new ThreadPoolExecutor(1, (int) (Runtime.getRuntime().maxMemory() * 0.1 / (1024 * 1024)), 60000, MILLISECONDS, new LinkedBlockingQueue<>());
```

You could also check the JVM arguments to see if the `-Xss` parameter is set and divide by that size argument rather than the default 1BM (`(1024 * 1024)`) in the example above

Reply:
I'm aware of the thread semantics; you're right, for IO bound tasks using the JVM threadpool is not the right idea ([virtual threads](https://openjdk.org/jeps/425) will help here) given that most FS implementations don't use non-blocking IO. However, I'm not convinced making the dynamic calculation is the right choice given that 1/We haven't done something similar in other bg threadpools 2/The connectors can (and do) use their own bg threadpools -this would our 10-15% memory usage assumption incorrect

I would rather use a (configurable) constant value for the max sizes. I think we should keep them conservative to start off
1. `100` for the QuickStatsBuilder - This would limit the max number of partitons scanned at any time to 100
2.  `500` for the ParquetQuickStats builder - Assuming 5 files per partition.

Of course these wont limit the actual partition/files that a query with quick stats can use since the bg builds will be queued up when we hit the limits. I will add threadpool monitoring metrics so that actual threadpool sizes can be monitored

Reply:
I'm fine setting to the default values you suggested.

My main reason for suggesting a dynamic default is because I've seen that once you start trying to scale to different hardware (up or down), it can take a fair bit of time to tune and debug which static configs need to be changed for a given deployment. A dynamic value as the default can potentially prevent the need to manually tune which makes the software more user friendly. I know it's generally not the norm with these types of systems, but I haven't seen strong arguments as to why we shouldn't be trying to tune some of the defaults dynamically based on the current hardware/JVM limits. More of a philosophical debate, but I would personally like to see things move in that direction